End-to-end BERT_pytorch working with dynamic shapes and inductor

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89313

This contains https://github.com/pytorch/pytorch/pull/88546 plus a big pile of inductor hacks.

```
$ TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python benchmarks/dynamo/torchbench.py --accuracy --backend inductor --training --only BERT_pytorch
cuda train BERT_pytorch                       PASS
```

I don't know if we're actually generating generic kernels though. Maybe @Chillee can check.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

ezyang committed:
End-to-end BERT_pytorch working with dynamic shapes and inductor

This contains https://github.com/pytorch/pytorch/pull/88546 plus
a big pile of inductor hacks.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

voznesenskym commented on functorch/_src/aot_autograd.py:
```diff
@@ -908,47 +942,44 @@ def functional_call(*args, **kwargs):
+        return aot_dispatcher_function(args)
 
-    compiled_f = aot_function_simplified(functional_call, *top_args, **top_kwargs)
+    # [Real vs Fake Params]
```
Hey, I missed this comment in my PR, but can you please add a comment saying to move this like:

`compiled_f = aot_function_simplified(functional_call, *top_args, **top_kwargs)`

Into def `forward` as a way of disabling the cache?

voznesenskym commented on torch/_dynamo/guards.py:
```diff
@@ -448,26 +449,34 @@ def SYMBOL_MATCH():
+        # STOP - DO NOT USE id_ref FOR TENSORS - TENSOR INVALIDATION RULES DIFFER
+        self.tensor_check_ids[tensor_name] = id(value)
+        if value._base is not None:
+            assert id(value._base) in self.guarded_code.output_graph.tensor_id_to_sym_shape_ref
+            refs = self.guarded_code.output_graph.tensor_id_to_sym_shape_ref[id(value._base)]
+            for ref in refs:
+                self.guarded_code.output_graph.base_symbols.update({ref.expr: f"{tensor_name}._base"})
```
this needs an `if config.dynamic_shapes` in front of it

ezyang committed:
Update on "End-to-end BERT_pytorch working with dynamic shapes and inductor"


This contains https://github.com/pytorch/pytorch/pull/88546 plus a big pile of inductor hacks.

```
$ TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python benchmarks/dynamo/torchbench.py --accuracy --backend inductor --training --only BERT_pytorch
cuda train BERT_pytorch                       PASS
```

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "End-to-end BERT_pytorch working with dynamic shapes and inductor"


This contains https://github.com/pytorch/pytorch/pull/88546 plus a big pile of inductor hacks.

```
$ TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python benchmarks/dynamo/torchbench.py --accuracy --backend inductor --training --only BERT_pytorch
cuda train BERT_pytorch                       PASS
```

I don't know if we're actually generating generic kernels though. Maybe Chillee can check.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "End-to-end BERT_pytorch working with dynamic shapes and inductor"


This contains https://github.com/pytorch/pytorch/pull/88546 plus a big pile of inductor hacks.

```
$ TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python benchmarks/dynamo/torchbench.py --accuracy --backend inductor --training --only BERT_pytorch
cuda train BERT_pytorch                       PASS
```

I don't know if we're actually generating generic kernels though. Maybe Chillee can check.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented:
I'm kind of thinking I don't really want to land this PR all in one go
